File: advance/src/eu/advance/logistics/flow/engine/api/ds/AdvanceKeyEntry.java
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */

package eu.advance.logistics.flow.engine.api.ds;

import hu.akarnokd.reactive4java.base.Func0;

import java.text.ParseException;
import java.util.Date;

import org.slf4j.LoggerFactory;

import eu.advance.logistics.flow.engine.api.core.Copyable;
import eu.advance.logistics.flow.engine.api.core.Identifiable;
import eu.advance.logistics.flow.engine.xml.XElement;
import eu.advance.logistics.flow.engine.xml.XSerializable;


/**
 * Properties of a concrete key in a key store.
 * @author akarnokd, 2011.09.20.
 */
public class AdvanceKeyEntry implements XSerializable, Copyable<AdvanceKeyEntry>, Identifiable<String> {
	/** The key type. */
	public AdvanceKeyType type;
	/** The key name. */
	public String name;
	/** The creation date. */
	public Date createdAt;
	/** Creates a new instance of this class. */
	public static final Func0<AdvanceKeyEntry> CREATOR = new Func0<AdvanceKeyEntry>() {
		@Override
		public AdvanceKeyEntry invoke() {
			return new AdvanceKeyEntry();
		}
	};
	@Override
	public AdvanceKeyEntry copy() {
		AdvanceKeyEntry result = new AdvanceKeyEntry();
		result.type = type;
		result.name = name;
		result.createdAt = new Date(createdAt.getTime());
		return result;
	}
	@Override
	public void load(XElement source) {
		type = AdvanceKeyType.valueOf(source.get("type"));
		name = source.get("name");
		try {
			createdAt = XElement.parseDateTime(source.get("created-at"));
		} catch (ParseException ex) {
			LoggerFactory.getLogger(AdvanceKeyEntry.class).error(ex.toString(), ex);
		}
	}
	@Override
	public void save(XElement destination) {
		destination.set("type", type);
		destination.set("name", name);
		destination.set("created-at", XElement.formatDateTime(createdAt));
	}
	@Override
	public String id() {
		return name;
	}
}

File: advance-api-cs/AdvanceAPIClient/Classes/AdvanceKeyEntry.cs
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Globalization;
using System.Xml;

using AdvanceAPIClient.Core;
using AdvanceAPIClient.Interfaces;

namespace AdvanceAPIClient.Classes
{
    /// <summary>
    /// Properties of a key in a key store.
    /// </summary>
    public class AdvanceKeyEntry : XmlReadWrite, ICopiable<AdvanceKeyEntry>, IIdentifiable<String>
    {
        /// <summary>
        /// Key type
        /// </summary>
        public AdvanceKeyType Type;
        /// <summary>
        /// Key name
        /// </summary>
        public string Id { get { return this.Name; } }
        public string Name;
        /// <summary>
        /// Creation date
        /// </summary>
        public DateTime CreatedAt;

        public AdvanceKeyEntry() : base() { }

        public AdvanceKeyEntry Copy()
        {
            AdvanceKeyEntry result = new AdvanceKeyEntry();
            result.Type = this.Type;
            result.Name = this.Name;
            result.CreatedAt = new DateTime(this.CreatedAt.Ticks);
            return result;
        }

        /// <summary>
        /// Fill object data from xml node
        /// </summary>
        /// <param name="source">Source Xml</param>
        protected override void LoadFromXmlNode(XmlNode source)
        {
            this.Type = GetEnumAttribute<AdvanceKeyType>(source, "type", AdvanceKeyType.PRIVATE_KEY);
            this.Name = GetAttribute(source, "name", null);
            this.CreatedAt = GetDateTimeAttribute(source, "created-at");
        }

        /// <summary>
        /// Fills  object's xml
        /// </summary>
        /// <param name="node">Empty node for the object with name "tagname"</param>
        protected override void FillXmlElement(XmlElement node)
        {
            AddAttribute(node, "type", this.Type);
            AddAttribute(node, "name", this.Name);
            AddDateTimeAttribute(node, "created-at", this.CreatedAt);
        }

    }
}
